src/OrbitGl/App.cpp

```diff
@@ -411,35 +411,67 @@ void OrbitApp::OnTimer(const TimerInfo& timer_info) {
-  // into TimerInfo registers with the Set method, the registers field of the copied TimerInfo
-  // would be empty.
-  std::vector<uint64_t> encoded_values(static_cast<size_t>(SystemMemoryUsageEncodingIndex::kEnd));
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kTotalKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.total_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kFreeKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.free_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kAvailableKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.available_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kBuffersKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.buffers_kb());
-  encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kCachedKb)] =
-      orbit_api::Encode<uint64_t>(system_memory_usage.cached_kb());
-
-  *timer_info.mutable_registers() = {encoded_values.begin(), encoded_values.end()};
-
-  GetMutableTimeGraph()->ProcessTimer(timer_info, nullptr);
+  if (memory_event_wrapper.has_system_memory_usage()) {
+    const orbit_grpc_protos::SystemMemoryUsage& system_memory_usage =
+        memory_event_wrapper.system_memory_usage();
+    TimerInfo timer_info;
+    timer_info.set_type(TimerInfo::kSystemMemoryUsage);
+    timer_info.set_start(memory_event_wrapper.timestamp_ns());
+    timer_info.set_end(memory_event_wrapper.timestamp_ns());
+
+    // The RepeatedField<Element>::Set(int index, const Element & value) method won't update the
+    // current_size_ of the repeated field. If the current_size_ == 0, the repeated field will be
+    // ignored when copying the protobuf. Therefore, if encoding the following memory information
+    // into TimerInfo registers with the Set method, the registers field of the copied TimerInfo
+    // would be empty.
+    std::vector<uint64_t> encoded_values(static_cast<size_t>(SystemMemoryUsageEncodingIndex::kEnd));
+    encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kTotalKb)] =
+        orbit_api::Encode<uint64_t>(system_memory_usage.total_kb());
+    encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kFreeKb)] =
+        orbit_api::Encode<uint64_t>(system_memory_usage.free_kb());
+    encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kAvailableKb)] =
+        orbit_api::Encode<uint64_t>(system_memory_usage.available_kb());
+    encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kBuffersKb)] =
+        orbit_api::Encode<uint64_t>(system_memory_usage.buffers_kb());
+    encoded_values[static_cast<size_t>(SystemMemoryUsageEncodingIndex::kCachedKb)] =
+        orbit_api::Encode<uint64_t>(system_memory_usage.cached_kb());
+
+    *timer_info.mutable_registers() = {encoded_values.begin(), encoded_values.end()};
```

Comment:
I'm confused here. `TimerInfo::registers`, which are specific to manual and dynamic instrumentation, are used to store memory data. And `orbit_api::Encode`, which is specific to manual instrumentation, is used as a tool for this. This is quite a hack. I think it should be implemented differently with dedicated fields.

Reply:
> I'm not sure whether it is a good idea to add all the fields into TimerInfo as they are used only for the memory tracks.

Your doubt comes from the fact that `TimerInfo` is a big mess, unfortunately. There are various fields, but most only makes sense in some cases, based on the `type` of `TimerInfo`. Considering the current state, I don't see a problem adding fields only used for memory tracks. I think that reusing fields that were introduced for other purposes (like `registers`) is worse.

You could wrap memory information into a sub-message `MemoryTimerInfo` (the name is just an example) of `TimerInfo`. Or you could do something similar to adding a repeated field of `VariableInfo` to `TimerInfo` as you suggested, but I would change `VariableInfo` to something that makes it clearer it's for memory, e.g., `MemoryValue` or something like that.

Now, all this said, I just noticed that `MemoryTrack` is not a `TimerTrack` (it inherits from `GraphTrack` which is just a `Track`). In that case, I think that you should just not use `TimerInfo`s at all, nor the `OnTimer` methods, it just makes things more complicated. I see that `GraphTrack` already uses `TimerInfo`s but I'm not sure of the point. In fact, you don't need to use protobufs at all (`TimerInfo` is a proto to load from the old capture format). You should probably just route instances of dedicated classes or structs to the `MemoryTrack`s.